timesync/peersync/sync.go

```diff
@@ -0,0 +1,330 @@
+type Sync struct {
+	errCnt uint32
+
+	config       Config
+	log          log.Log
+	srv          *server.MessageServer
+	time         Time
+	network      Network
+	peersWatcher peersWatcher
+
+	once   sync.Once
+	wg     *errgroup.Group
+	ctx    context.Context
+	cancel func()
+}
+
+func (s *Sync) requestHandler(ctx context.Context, buf []byte) []byte {
+	var request Request
+	if err := types.BytesToInterface(buf, &request); err != nil {
+		s.log.Debug("can't decode request", log.Binary("request", buf), log.Err(err))
+		return nil
+	}
+	resp := Response{
+		ID:        request.ID,
+		Timestamp: s.time.Now().UnixNano(),
+	}
+	buf, err := types.InterfaceToBytes(&resp)
+	if err != nil {
+		s.log.Panic("can't encode response", log.Binary("response", buf), log.Err(err))
+	}
+	return buf
+}
+
+// Start background workers.
+func (s *Sync) Start() {
+	s.once.Do(func() {
+		s.wg.Go(func() error {
+			return s.run()
+		})
+		s.wg.Go(func() error {
+			added, expired := s.network.SubscribePeerEvents()
+			return s.peersWatcher.run(s.ctx, added, expired)
+		})
+	})
+}
+
+// Stop background workers.
+func (s *Sync) Stop() {
+	s.cancel()
+	s.srv.Close()
+	s.Wait()
+}
+
+// Wait will return first error that is returned by background workers.
+func (s *Sync) Wait() error {
+	err := s.wg.Wait()
+	if errors.Is(err, context.Canceled) {
+		return nil
+	}
+	return err
+}
+
+func (s *Sync) run() error {
+	var (
+		timer *time.Timer
+		round uint64
+	)
+	s.log.With().Debug("started sync background worker")
+	defer s.log.With().Debug("exiting sync background worker")
+	for {
+		prs, err := s.peersWatcher.waitPeers(s.ctx, s.config.RequiredResponses)
+		if err != nil {
+			return err
+		}
+		s.log.With().Info("starting time sync round with peers",
+			log.Uint64("round", round),
+			log.Int("peers_count", len(prs)),
+			log.Uint32("errors_count", atomic.LoadUint32(&s.errCnt)),
+		)
+		rctx, cancel := context.WithTimeout(s.ctx, s.config.RoundTimeout)
+		offset, err := s.GetOffset(rctx, round, prs)
+		cancel()
+
+		var timeout time.Duration
+		if err == nil {
+			if offset > s.config.MaxClockOffset || (offset < 0 && -offset > s.config.MaxClockOffset) {
+				s.log.With().Error("peers offset is larger than max allowed clock difference",
+					log.Uint64("round", round),
+					log.Duration("offset", offset),
+					log.Duration("max_offset", s.config.MaxClockOffset),
+				)
+				if atomic.AddUint32(&s.errCnt, 1) == uint32(s.config.MaxOffsetErrors) {
+					return ErrPeersNotSynced
+				}
+			} else {
+				s.log.With().Info("peers offset is within max allowed clock difference",
+					log.Uint64("round", round),
+					log.Duration("offset", offset),
+					log.Duration("max_offset", s.config.MaxClockOffset),
+				)
+				atomic.StoreUint32(&s.errCnt, 0)
+			}
+			timeout = s.config.RoundInterval
+		} else {
+			s.log.With().Error("failed to fetch offset from peers", log.Err(err))
+			timeout = s.config.RoundRetryInterval
+		}
+
+		round++
```

Comment:
why do we need round counter?

Reply:
initially i needed it to avoid counting delayed votes. i know that msgServer will already prevent it, as it have the unique id for each request. but decided to keep it anyway so that protocol can work over any connection without req/response server. can remove it if you prefer it